portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -16,49 +16,69 @@
 
 import com.google.sps.data.Comment;
 import com.google.gson.Gson;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import java.util.ArrayList;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that returns and saves comments in Datastore */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-
-  private ArrayList<Comment> commentList;
-
-  @Override
-  public void init() {
-    commentList = new ArrayList<Comment>();
-  }
   
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    json = convertToJson(commentList);
+    int commentLimit = Integer.parseInt(request.getParameter("limit"));
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    ArrayList<Comment> commentList = new ArrayList<Comment>();
+    for (Entity entity : results.asIterable()) {
+      long id = entity.getKey().getId();
+      String name = (String) entity.getProperty("name");
+      String comment = (String) entity.getProperty("comment");
+      long timestamp = (long) entity.getProperty("timestamp");
+
+      Comment commentObject = new Comment(id, name, comment, timestamp);
+      commentList.add(commentObject);
+    }
+    
+    Gson gson = new Gson();   
+    String json;
+    if (commentLimit < commentList.size()) {
+      json = gson.toJson(commentList.subList(0,commentLimit));
```

Comment:
You can keep tracking the size of commentList in the above for loop, and exit early once it reaches to the limit.

Reply:
Thanks for the suggestion but I no longer need to track of the size of commentList because I recently found a way to set a limit for the query to control the maximum number of results returned. I committed the change I made.
https://cloud.google.com/appengine/docs/standard/java/datastore/retrieving-query-results#setting_limits